pkg/auth/kubesecret.go

```diff
@@ -0,0 +1,175 @@
+			return
+		}
+		if err := listener.SyncKubeSecrets(ctx, clientset); err != nil {
+			logrus.WithError(err).Infof("failed to sync secrets")
+			return
+		}
+	}()
+	return listener
+}
+
+func (kubelistener *KubeSecretListener) SyncKubeSecrets(ctx context.Context, clientset *kubernetes.Clientset) error {
+	if kubelistener.informer == nil {
+		informer := cache.NewSharedIndexInformer(
+			&cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+					options.FieldSelector = "type=" + string(corev1.SecretTypeDockerConfigJson)
+					return clientset.CoreV1().Secrets(metav1.NamespaceAll).List(context.Background(), options)
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					options.FieldSelector = "type=" + string(corev1.SecretTypeDockerConfigJson)
+					return clientset.CoreV1().Secrets(metav1.NamespaceAll).Watch(context.Background(), options)
+				}},
+			&corev1.Secret{},
+			0,
+			cache.Indexers{},
+		)
+		kubelistener.informer = informer
+		kubelistener.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err != nil {
+					logrus.WithError(err).Errorf("failed to get key for secret from cache")
+					return
+				}
+				data, ok := obj.(*corev1.Secret).Data[corev1.DockerConfigJsonKey]
+				if !ok {
+					logrus.Infof("failed to get data from secret")
+					return
+				}
+				dockerConfig := configfile.ConfigFile{}
+				err = dockerConfig.LoadFromReader(bytes.NewReader(data))
+				if err != nil {
+					logrus.WithError(err).Infof("failed to load docker config json from secret")
+					return
+				}
+				kubelistener.configMu.Lock()
+				kubelistener.dockerConfigs[key] = &dockerConfig
+				kubelistener.configMu.Unlock()
+			},
+			UpdateFunc: func(old, new interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(new)
+				if err != nil {
+					logrus.WithError(err).Errorf("failed to get key for secret from cache")
+					return
+				}
+				data, ok := new.(*corev1.Secret).Data[corev1.DockerConfigJsonKey]
+				if !ok {
+					logrus.Infof("failed to get data from secret")
+					return
+				}
+				dockerConfig := configfile.ConfigFile{}
+				err = dockerConfig.LoadFromReader(bytes.NewReader(data))
+				if err != nil {
+					logrus.WithError(err).Infof("failed to load docker config json from secre")
+					return
+				}
+				kubelistener.configMu.Lock()
+				kubelistener.dockerConfigs[key] = &dockerConfig
+				kubelistener.configMu.Unlock()
+			},
+			DeleteFunc: func(obj interface{}) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err != nil {
+					logrus.WithError(err).Errorf("failed to get key for secret from cache")
+					return
+				}
+				kubelistener.configMu.Lock()
+				delete(kubelistener.dockerConfigs, key)
+				kubelistener.configMu.Unlock()
+			}},
+		)
+		go kubelistener.informer.Run(ctx.Done())
+		if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
```

Comment:
Will this watch secrets until when?

Reply:
The informer will be stopped when stopCh from ctx.Done()  is closed